workerpool/pool.go

```diff
@@ -58,20 +63,41 @@ func (pool *WorkerPool) Start() error {
 				for _, worker := range job.workers {
 					worker.done <- true
 				}
+
+				if err := job.mutex.Unlock(context.TODO()); err != nil {
+					pool.log.Errorf("unable to unlock", "error", err, "connectionID", job.connection.ID)
+				}
+				pool.log.Debugf("unlocked", "connectionID", job.connection.ID)
 			}
 			return nil
 		case event := <-pool.events:
-			count := event.Connection.Source.NumberOfWorkers()
-			if _, ok := pool.jobs[event.Connection.ID]; !ok {
-				pool.jobs[event.Connection.ID] = &job{
-					connection: event.Connection,
-					workers:    make(map[uint]*worker),
+			if event.Type == watcher.Created {
+				// acquire lock
+				mutex := concurrency.NewMutex(pool.session, lockPrefix+string(event.ID))
+				pool.log.Debugw("locking...", "connectionID", event.ID)
+				ctx, cancel := context.WithTimeout(context.Background(), time.Second*2)
+				err := mutex.Lock(ctx)
```

Comment:
I would swap these two lines... and make it 

```
ctx, cancel := context.WithTimeout(context.Background(), time.Second * 2)
defer cancel()
if err := mutex.Lock(ctx); err != nil {
    	pool.log.Debugw("unable to lock", "error", err, "connectionID", event.ID)
	    continue
}
```

Reply:
It cannot be like that because `cancel` has to be called after `Lock`.